### ISSUE TITLE
Extremely slow page turns

### ISSUE THREAD

**Issue by polo2137** [koreader/koreader#12119]
* KOReader version: 202404000209
* Device: Kobo Libra 2
* Font: OpenDyslexic

Page turns in a epub book with a large amount of highlights in different styles take very long. Crashlog at the end of the post.
https://github.com/koreader/koreader/assets/123873769/59acd5be-35b5-4f91-addf-74ea63892162

Bonus bug:
Uneven highlights after dashes.
![IMG_5011](https://github.com/koreader/koreader/assets/123873769/b893149b-df74-49a8-a3e8-4d724c19299c)

[crash.log](https://github.com/user-attachments/files/16034868/crash.log)

**Comment by poire-z** [comment]
In the crash.log, there are mostly a lot of:
```
06/27/24-00:46:27 INFO  Wi-Fi successfully restored (after 0.25 seconds)!
./luajit: frontend/ui/otamanager.lua:371: attempt to call method 'getZsyncFilename' (a nil value)
stack traceback:
    frontend/ui/otamanager.lua:371: in function 'zsync'
    frontend/ui/otamanager.lua:229: in function 'action'
    frontend/ui/uimanager.lua:1020: in function '_checkTasks'
    frontend/ui/uimanager.lua:1470: in function 'handleInput'
    frontend/ui/uimanager.lua:1590: in function 'run'
    ./reader.lua:280: in main chunk
    [C]: at 0x000140cd
!!!!
Uh oh, something went awry... (Crash nÂ°1: 06/27/24 @ 00:46:32)
```
which has been reported by @mergen3107 in #12115.

About the slow page turns, dunno, @hius07 may be able to tell if that's expected (if it is a xpointers comparisons/loops, or if drawing is expensive).
(Also, be sure you have in `File browser > Tools > More tools > Developer options > {x] Enable cre call cache` really checked.)

> Bonus bug:
> Uneven highlights after dashes.

Not sure it's a bug.
I don't think you have "Typography rules > [x] Hanging punctuation" checked (or you would also have the comma and period highlights go a bit out too - but check again: if you disable that, everything should be even:
![image](https://github.com/koreader/koreader/assets/24273478/951b5454-dd2b-4db3-8ed3-1dfcb72af08f)

It's possible it is this specific font, which may have the hyphen glyph a bit strange, ie. the ink of the bitmap overflowing the "blackbox" (the advertized size of the glyph, that we use for sizing and positionning glyphs and adjusting words position). May be they use this trick to actually expect such hyphens to hang in the margin (which is what our `[x] Hanging punctuation` is doing (because it's nicer fot reading - but not nicer with highlights), expecting the fonts to not do anything).
See if this happens with other fonts.

**Comment by polo2137** [comment]
Update problem:
After reviewing the issue #12115 I still really don't know how to fix this update problem. Sorry I really can't use commands etc. only GUI unless someone explain me step by step how to do certain thing. I encountered said issue while trying to downgrade to a stable version. Does that mean I will not be able to update koreader anymore without using commands?

Slow pages turn problem:
CRE call cache is enabled.

Font problem:
Other fonts indeed are aligned correctly (solved).

**Comment by mergen3107** [comment]
@polo2137 
For versions, you can just get another koreader build and override your `koreader` folder. Wait until the OTA issue is fixed and then update.

**Comment by polo2137** [comment]
@mergen3107 How to override koreader folder on mac?

**Comment by mergen3107** [comment]
Builds don’t come with books or settings, you can just overwrite

**Comment by polo2137** [comment]
@mergen3107 When I drag koreader folder to .adds and then merge them my koreader version is still the same developer beta.

**Comment by mergen3107** [comment]
Well, of course! The fix is not there yet :D 

> Wait until the OTA issue is fixed and then update.

**Comment by polo2137** [comment]
Ohh, right sorry then and thanks!

The issue with slow page turns still persists so if anyone would be nice enough to look into it I will be extremely glad!

**Comment by poire-z** [comment]
```
06/27/24-00:57:49 INFO  Inhibiting user input
06/27/24-00:57:50 INFO  Restoring user input handling
06/27/24-01:03:22 INFO  Inhibiting user input
06/27/24-01:03:23 INFO  Restoring user input handling
06/27/24-01:03:37 INFO  Inhibiting user input
06/27/24-01:03:38 INFO  Restoring user input handling
06/27/24-01:03:43 INFO  Inhibiting user input
06/27/24-01:03:44 INFO  Restoring user input handling
06/27/24-01:03:47 INFO  Inhibiting user input
06/27/24-01:03:48 INFO  Restoring user input handling
06/27/24-01:03:50 INFO  Inhibiting user input
06/27/24-01:03:51 INFO  Restoring user input handling
06/27/24-01:03:52 INFO  Inhibiting user input
06/27/24-01:03:53 INFO  Restoring user input handling
06/27/24-01:03:54 INFO  Inhibiting user input
06/27/24-01:03:54 INFO  Restoring user input handling
06/27/24-01:03:56 INFO  Inhibiting user input
06/27/24-01:03:57 INFO  Restoring user input handling
06/27/24-01:04:25 INFO  Inhibiting user input
06/27/24-01:04:26 INFO  Restoring user input handling
06/27/24-01:04:31 INFO  Inhibiting user input
06/27/24-01:04:32 INFO  Restoring user input handling
06/27/24-01:04:47 INFO  Inhibiting user input
06/27/24-01:04:48 INFO  Restoring user input handling
06/27/24-01:04:52 INFO  Inhibiting user input
06/27/24-01:04:53 INFO  Restoring user input handling
06/27/24-01:04:58 INFO  Inhibiting user input
06/27/24-01:04:59 INFO  Restoring user input handling
06/27/24-01:05:04 INFO  Inhibiting user input
06/27/24-01:05:05 INFO  Restoring user input handling
```

Is that you turning pages?
Dunno where these come from...
(If you have some auto standby plugin ? Try disabling it, may be it's the wakeup on page turn.)
No issue on that book before you made highlights ?

You could try enabling debug mode (in developper options) for a few seconds and a few page turns, then disable it, and give us the few hundreds of lines with DEBUG lines, so we see what's between these INFO lines.

**Comment by polo2137** [comment]
I had auto standby plugin enabled in plugin menu but not actually enabled it to take effect.
There was no issue before making highlights.

Here is the crashlog with enabled developer logging.
[crash.log](https://github.com/user-attachments/files/16037040/crash.log)

**Comment by mergen3107** [comment]
(Restart after renaming)

**Comment by polo2137** [comment]
>Rename koreader/data/dict to dict2 (temporarily) and see if issue persists.

No change :/

**Comment by mergen3107** [comment]
Are you still seeing these lines with dict renamed?

Cleaner verbose crashlog would be nice :D

**Comment by polo2137** [comment]
There you go :D
[crash.log](https://github.com/user-attachments/files/16037180/crash.log)

**Comment by mergen3107** [comment]
Interesting, I can see touch events as duplicates (2-6 times) per timestamp.

**Comment by polo2137** [comment]
I assure you that for every page turned there is just one button click or screen swipe on my end :)

**Comment by polo2137** [comment]
Also only one page at a time is turned.

**Comment by poire-z** [comment]
I don't see anything really slow, all page turns things happen in the same second.
May be try to disable the top status bar (we tweaked the page info display recently) and see if it is better.

**Comment by polo2137** [comment]
>  May be try to disable the top status bar (we tweaked the page info display recently) and see if it is better.

Thats it!!! Thanks so much!!!
Every option referencing to book progress in alt status bar does slow page turns: current page, total pages and progress percentage.

**Comment by poire-z** [comment]
It's still something we need to look at then.
And what about the footer? Do you have it enabled, showing the same kind of page progress info ? And it's not slow when you only have the footer?
And check with another book of approximately the same size where you don't have any highlight. Also slow with the top status bar enabled?
When (on what day) did you notice it became slower ?

**Comment by poire-z** [comment]
I can reproduce with our "Human action" epub with many highlights:
- no slowness with the footer
- not slow with the header when there are no highlights
- very slow with the header when there are hundreds of highlights

**Comment by polo2137** [comment]
> And what about the footer? Do you have it enabled, showing the same kind of page progress info ?
Yes it's enabled with the same kind of book progress info.
> And it's not slow when you only have the footer?
No, it's alright.
> And check with another book of approximately the same size where you don't have any highlight. Also slow with the top status bar enabled?
No, all works just fine. Everything is working well even with a little number (3) of highlights.
> When (on what day) did you notice it became slower ?
After installing newest developer beta.

**Comment by poire-z** [comment]
And it happens since my #11873 (so, not cause by the few other tweaks & fixes after it).

**Comment by polo2137** [comment]
So maybe it's just a coincidence that it started happening after installing beta version. Probably after installing it I just highlighted enough words to make this bug happen.

**Comment by poire-z** [comment]
Yes, coincidence - but thank you for reporting it.
Should be fixed by the one line added in #12123.
(You can try adding it and re-enabling the top status bar, and hopefully report all is fine now.)

**Comment by polo2137** [comment]
>(You can try adding it and re-enabling the top status bar, and hopefully report all is fine now.)
Can confirm it works now.
